### PR TITLE
fix(gatsby): fix typings in resolve-module-exports

### PR DESCRIPTION
**Pull request by wardpeet**
# Description

When I rebased #25233 to master I'm getting type errors. I don't know how this slipped through the nets but at least we should do a full typecheck on master.

I also fixed typing in that file.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -212,7 +212,13 @@ jobs:
     executor: node
     steps:
       - checkout
-      - run: ./scripts/assert-changed-files.sh "packages/*|.circleci/*"
+      ## We should always assert on master
```
Comment (LekoArts) [inline .circleci/config.yml:215]:
```suggestion
      # We should always assert on master
```

typo

File: packages/gatsby/src/bootstrap/resolve-module-exports.ts [hunk 2]
```diff
@@ -81,15 +81,30 @@ const staticallyAnalyzeExports = (
...
+      const exportName = astPath?.node?.exported?.name
       isES6 = true
-      if (exportName) exportNames.push(exportName)
+      if (exportName) {
+        exportNames.push(exportName)
+      }
     },
 
     // export default () => {}
     // const foo = () => {}; export default foo
```
Comment (LekoArts) [inline packages/gatsby/src/bootstrap/resolve-module-exports.ts:92]: Can you also add the added asserted functions to the comments here?

File: packages/gatsby/src/bootstrap/resolve-module-exports.ts [hunk 3]
```diff
@@ -81,15 +81,32 @@ const staticallyAnalyzeExports = (
...
+    // export default function() {}
+    // export default function foo() {}
     // const foo = () => {}; export default foo
     ExportDefaultDeclaration: function ExportDefaultDeclaration(astPath) {
-      const name = get(astPath, `node.declaration.name`)
+      const declaration = astPath.node.declaration
+      if (
+        !t.isIdentifier(declaration) &&
+        !t.isArrowFunctionExpression(declaration) &&
+        !t.isFunctionDeclaration(declaration)
```
Comment (pvdz) [inline packages/gatsby/src/bootstrap/resolve-module-exports.ts:100]: Is it not relevant to check the name of a function declaration? (In that case do note a notorious exception where default exported function declarations might not have an explicit name.)

Also, it's very technically possible that this is a function expression (`export default (function(){})`), which may be less of an edge case with TS type casting or something. Is that relevant here? Those functions may or may not have an id as well.

Comment (wardpeet) [reply packages/gatsby/src/bootstrap/resolve-module-exports.ts:100]: There are lot of things we don't account for but mosty I just want to unblock myself. I've added the function identifier to mimic arrow functions.

File: .circleci/config.yml [hunk 4]
```diff
@@ -212,7 +212,13 @@ jobs:
     executor: node
     steps:
       - checkout
-      - run: ./scripts/assert-changed-files.sh "packages/*|.circleci/*"
+      # We should always assert on master
+      - when:
+          condition:
+            - not:
+                equal: [master, << pipeline.git.branch >>]
```
Comment (blainekasten) [inline .circleci/config.yml:219]: can you explain what this is doing?

Comment (wardpeet) [reply .circleci/config.yml:219]: always run this on master, else use the packages detection skip

**Pull request merged**